compiler/rustc_type_ir/src/fast_reject.rs

```diff
@@ -214,184 +211,183 @@ impl<I: Interner> DeepRejectCtxt<I> {
-            | ty::Dynamic(..)
-            | ty::Pat(..)
-            | ty::Ref(..)
-            | ty::Never
-            | ty::Tuple(..)
-            | ty::FnPtr(..)
-            | ty::Foreign(..) => debug_assert!(impl_ty.is_known_rigid()),
-            ty::FnDef(..)
-            | ty::Closure(..)
-            | ty::CoroutineClosure(..)
-            | ty::Coroutine(..)
-            | ty::CoroutineWitness(..)
-            | ty::Placeholder(..)
-            | ty::Bound(..)
-            | ty::Infer(_) => panic!("unexpected impl_ty: {impl_ty:?}"),
-        }
+    pub fn types_may_unify(self, lhs: I::Ty, rhs: I::Ty) -> bool {
+        match (lhs.kind(), rhs.kind()) {
+            (ty::Error(_), _) | (_, ty::Error(_)) => true,
 
-        let k = impl_ty.kind();
-        match obligation_ty.kind() {
-            // Purely rigid types, use structural equivalence.
-            ty::Bool
-            | ty::Char
-            | ty::Int(_)
-            | ty::Uint(_)
-            | ty::Float(_)
-            | ty::Str
-            | ty::Never
-            | ty::Foreign(_) => obligation_ty == impl_ty,
-            ty::Ref(_, obl_ty, obl_mutbl) => match k {
-                ty::Ref(_, impl_ty, impl_mutbl) => {
-                    obl_mutbl == impl_mutbl && self.types_may_unify(obl_ty, impl_ty)
-                }
-                _ => false,
-            },
-            ty::Adt(obl_def, obl_args) => match k {
-                ty::Adt(impl_def, impl_args) => {
-                    obl_def == impl_def && self.args_may_unify(obl_args, impl_args)
+            // As we're walking the whole type, it may encounter projections
+            // inside of binders and what not, so we're just going to assume that
+            // projections can unify with other stuff.
+            //
+            // Looking forward to lazy normalization this is the safer strategy anyways.
+            (ty::Alias(..), _) | (_, ty::Alias(..)) => true,
+
+            // Bound type variables may unify with rigid types e.g. when using
+            // non-lifetime binders.
+            (ty::Bound(..), _) | (_, ty::Bound(..)) => true,
+
+            (ty::Infer(var), _) => self.var_and_ty_may_unify(var, rhs),
+            (_, ty::Infer(var)) => self.var_and_ty_may_unify(var, lhs),
+
+            (ty::Param(lhs), ty::Param(rhs)) => {
+                match (self.treat_lhs_params, self.treat_rhs_params) {
+                    (TreatParams::AsRigid, TreatParams::AsRigid) => lhs == rhs,
+                    (TreatParams::InstantiateWithInfer, TreatParams::AsRigid)
+                    | (TreatParams::AsRigid, TreatParams::InstantiateWithInfer)
+                    | (TreatParams::InstantiateWithInfer, TreatParams::InstantiateWithInfer) => {
+                        true
+                    }
                 }
-                _ => false,
-            },
-            ty::Pat(obl_ty, _) => {
+            }
+            (ty::Param(_), ty::Placeholder(_)) | (ty::Placeholder(_), ty::Param(_)) => true,
```

Comment:
not really relevant as it doesnt ever happen on stable, but `ty::Param` and placeholder can't unify when treating params as rigid, so you can remove this branch and just use the `self.treat_lhs_params == TreatParams::InstantiateWithInfer` branch